### DISCUSSION TITLE
Naming/structure conventions for 1962 Calendar

### DISCUSSION THREAD

**Discussion by TobiTenno** [romcal/romcal#473] [Ideas]
I'm wanting to use what's been built to create a base calendar ala general-roman for the 1962 calendar. 

Technically it's the General Roman Calendar of 1960.

I'm currently working on the inputs for it to be used like a normal regional calendar, but given some feasts are Xth Sunday after Y, I'd still have some work to recognize those properly.

Questions:
- how would such a naming for this calendar work?
- would a change like this be welcome here?

Looks like there's already an open issue [here](https://github.com/romcal/romcal/issues/23)

**Comment by tukusejssirs** [comment]
@TobiTenno, Etienne repeatedly rejected the idea of adding support of calendars for historical revisions of the GRC (note: I usually call these calendars as _historical calendars_ in short, however, this term also includes historical versions of a revision, like when a celebration rank/date/name changes, a new new revision version is created). As you have noted, we have discussed this esp in #23.

I am still looking forward to supporting such calendars (even if just for educational purposes, esp calendar revisions before `cal1962`); however, the scope of the 1969 GRC revision (`cal1969`/`grc1969` in short) is quite huge. That said, I also think that romcal could support calendars of other churches in full communion with the Pope (instead of creating _greekcal_ or something like that).

Now, I think we should first refactor the repo into monorepo so that there would be no specific calendar revision logic in `@romcal/core` at all (although it might contain the Easter date calculations, however, those very from a revision/rite to another, thus it might be moved into a library which might be only internal/non-publishable). See [this comment of mine](https://github.com/romcal/romcal/discussions/445#discussioncomment-7262559), but note that it is a bit outdated; see mostly the notes on `@romcal/core` and `@romcal/grc1969/calendars`.

Therefore, in the end, IMHO the monorepo creation is the most important step IMHO. Here, we have two possible ways to do it:

1. move the current library version to a monorepo as it is, the re-structure romcal in smaller iterations:
   - while it is feasable doing so, it has a few issues:
      - the build script is quite a beast IMHO;
      - it won’t solve any underlining issues;
      - ...;
3. create a new version of romcal to fully re-structure romcal into (internal and publishable) libraries:
    - we would re-structure romcal into multiple (internal and publishable) libraries:
    - this would have a potential to solve all issues;
    - suggested rules: #528.

Note that while I am for the second option, although it requires more dev time, however, in the end IMHO, it would solve the data-related issues much quicker. (By the _data-related issues_, I mean issues regarding l10n, celebrations inscription in calendars, missing source references, and esp the complexity of adding, verifying, reviewing and merging the data. Solution: use a source/calendar/l10n tools to add the data. We would also move the data to separate packages, and import them from JSON/YAML files.)

**Reply by emagnier** [reply]
Transforming this project into a monorepo won't magically make Romcal compatible with any rite/missal//ordo.

Supporting another missal is not that simple. It's not just about generating the proper liturgical time. Romcal's core functionalities are also crucial for integrating and managing the many specific rules of the 1969 missal. Some rites, like the Ambrosian rite, are quite similar and could easily have a Romcal variant tailored for this rite (in fact, I explored this possibility, albeit not fully realized, in a modular re-architecture of Romcal in an exploratory git branch: `optimizations-ex`).

As for the older form of the Roman rite, I'm certainly not against an implementation (I'd actually think it would be great to offer it!), but I believe there are too many rule differences that are simply not applicable between the 2 missals, risking to complicate the Romcal core library (whose functionalities were developed specifically and closely following the rules of the 1969 missal). Additionally, there would likely need to be dedicated calendar definitions for each missal.

Ultimately, we can indeed share a lot of metadata about the blessed and saints. However, for the rest, it seems wise and reasonable not to mix tools for generating calendars specific to each missal, which have too many different rules (thus remaining two distinct libraries).

So far, I haven't been convinced by proposed architectures (not only data-oriented but also functional and "business" oriented). Therefore, feel free to propose POCs or implementation ideas.

The goals to keep in mind, for me, are to not mix the rules of the missals and to keep tools increasingly simple to use, maintain, and contribute to.

**Reply by TobiTenno** [reply]
I agree that all of them probably shouldn't live in the core library, but being open to them living in the logic for a particular calendar would be a big step, and be open to allowing a calendar to essentially define an expansion to allow derivations of the 1960 GRC to be extended as well would open up to being able progressively create revisions or variations based on those calendars "closest", like we do now for variant calendars

**Reply by tukusejssirs** [reply]
Note: although I react to the Etienne’s comment exclusively, I expect reactions from both of you. :wink:

> [@emagnier]
>
> Transforming this project into a monorepo won't magically make Romcal compatible with any rite/missal//ordo.
>
> Supporting another missal is not that simple. It's not just about generating the proper liturgical time. Romcal's core functionalities are also crucial for integrating and managing the many specific rules of the 1969 missal. Some rites, like the Ambrosian rite, are quite similar and could easily have a Romcal variant tailored for this rite (in fact, I explored this possibility, albeit not fully realized, in a modular re-architecture of Romcal in an exploratory git branch: `optimizations-ex`).

Etienne, IMHO `@romcal/core` should not contain any calendar computation logic. It should be an abstract/base/shell package which (based on its configuration) use the data from a particular GRC revision or rite (or the revision of the rite). Common stuff should be moved to internal (non-publishable) libraries (e.g. the Easter computus in both the Gregorian and Julian calendar). Then, we should publish a separate ‘core’ libraries for each calendar revision or rite, e.g. `@romcal/core/roman1962`, `@romcal/core/roman1969` or something along the lines, but no `@romcal/core` package would be published then (there would be a `core` package in the repo though). In a way, similar calendars could be based on / derived from each other if we want to make it more DRY (however, using internal libraries should be preferred even for this use case IMHO).

The actual logic would be in `@romcal/roman1969/general`, `@romcal/roman1962/general` where we would define all the rules for the Proper of Time and Proper of Saints (or similar for other revisions/rites).

@emagnier, @TobiTenno, WDYT?

> [@emagnier]
>
> As for the older form of the Roman rite, I'm certainly not against an implementation (I'd actually think it would be great to offer it!)

Sorry than for saying you are against it, however, that was my feeling from our older discussions.

> [@emagnier]
>
> Ultimately, we can indeed share a lot of metadata about the blessed and saints.

While this is mostly true, some celebration names/ranks/titles and liturgical seasons are different, however, that could be solved by adding additional properties to the celebration metadata and l10n files.

> [@emagnier]
>
> The goals to keep in mind, for me, are to not mix the rules of the missals and to keep tools increasingly simple to use, maintain, and contribute to.

I agree with you on this. We could copy this as a rule (possibly accepted by all of us?) to the OP of #528, WDYT?

**Reply by TobiTenno** [reply]
I think whatever the current in-force Ordinary for the Roman Catholic Church should be `general`, so the right now, even if we maintain a `@romcal/calendar1969.general-roman` it should be aliased to the current `@romcal/calendar.general-roman`

**Reply by emagnier** [reply]
@tukusejssirs, I might be mistaken, but from your reading, it seems like you may not realize all the rules and logic that are necessary and specific to creating a liturgical calendar for a particular missal/rite (I invite you to look closely at the implementation to see this). It appears that you are placing a lot of emphasis on "data architecture" or package names, but those aspects alone are not enough in this context.

As I mentioned above, I'm not opposed to supporting the 1962 missal. However, since there are many different rules to consider for generating this liturgical calendar base (especially the classification system of celebrations), upon which all liturgical days must be calculated, I strongly believe that this logic should be handled in a dedicated manner (because the contracts and "business logic" are different).
So let's try it experimentally, in a new separate module (let's transform this repository into a monorepo and add that here!). This will allow us to see to what extent such a project can be integrated, or not, into the current library and data.

Indeed, many other elements could be shared (calculation of Easter date, various data, etc.).

**Reply by tukusejssirs** [reply]
> I think whatever the current in-force Ordinary for the Roman Catholic Church should be `general`, so the right now, even if we maintain a `@romcal/calendar1969.general-roman` it should be aliased to the current `@romcal/calendar.general-roman`

I’d agree with using `general` the way you suggest, until we add other rites into the equation.

Therefore, I think even for the GRC revision, we should append the year of their promulgation. The prefix could be either `general` or `grc` (e.g. `general1969` or `general_1969` or `grc1969` or `grc_1969`). Or we should re-order the name parts differently, e.g. `@romcal/calendar/1969/general`/`@romcal/calendar/1969/general-roman` (i.e. the format could be `@romcal/calendar/${year}/${calendarName}`; we might need to add the rite too). (Maybe we need to replace the extra slashes with dots.)

In the end, we _could_ create some aliases, however, it could create a bit of extra work and might confuse the users. I’d still prefer an exact name over an inexact; OFC we need to define the naming rules and publish them in the `docs` app.

All in all, I don’t really care about the name, as long as it precisely specifies what the calendar is (i.e. it is a `calendar` promulgated in `1969` which is used in a specific place or is the GRC, possibly in a specific rite). At least that is my preference, however, I am open for a discussion.

**Reply by TobiTenno** [reply]
I'd avoided using directory dividers, because i think you can only have one, specifically right after scope, so we'd have to have other separators.

**Reply by tukusejssirs** [reply]
> [@emagnier]
>
> @tukusejssirs, I might be mistaken, but from your reading, it seems like you may not realize all the rules and logic that are necessary and specific to creating a liturgical calendar for a particular missal/rite (I invite you to look closely at the implementation to see this). It appears that you are placing a lot of emphasis on "data architecture" or package names, but those aspects alone are not enough in this context.

While I agree with you that I _plac[e] a lot of emphasis on "data architecture" and package names_ (I do that because I find the data to be a very important part of romcal, and through the years, I miss much metadata and a localisation tool which would ease the contribution/verification/review of the data), I think I understand that calendar revisions and calendars for different rites usually differ substantially. However, the data is ATM more important to me, as (1) I’d like to have them all sourced (and thus verified); and (2) I wish romcal would provide more celebration metadata that it currently provides. That said, the rules and logic are very important too, however, from my POV these are in better state than the data.

However, I see no reason why we could not move the differences (or even the all rules and whole logic) into separate (sub-) libraries (e.g. `@romcal/core/$revision_or_rite`) which would be selected based on the romcal configuration a user would provide (e.g. they could define what calendar revision or rite). The base library (e.g. `@romcal/core`) would load the correct library using which the correct celebration classification/names/ranks would be used and a calendar would be generated for a particular year for a specific calendar revision or rite. What problems do you see with this approach? :thinking:

Moreover, no one said romcal would have different GRC revisions and rites from day one. All we (or at least I) are talking about is to support them so that we could tell potential contributors either how to implement them or what information (rules, logic, data) we need to implement them.

All in all, I think we can agree that generation of calendars for different revisions/rites in a _single_ library is not the way to do it, however, moving the specific rules and logic to separate, internal libraries could be the way IMHO.

**Reply by tukusejssirs** [reply]
> [@TobiTenno]
>
> I'd avoided using directory dividers, because i think you can only have one, specifically right after scope, so we'd have to have other separators.

Yeah, I slightly remembered such a rule (I believe it is an NPM rule), that’s why I noted it in the brackets. :wink:

**Reply by emagnier** [reply]
@tukusejssirs, I think overall we're in agreement. To avoid getting bogged down in too many details here, let's proceed step by step, and solutions will naturally present themselves at the right time.
